pkg/ucp/store/postgres/postgresclient.go

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must refer to a named resource, not a collection"}
+	}
+
+	converted, err := storeutil.ConvertScopeIDToResourceID(parsed)
+	if err != nil {
+		return err
+	}
+
+	config := store.NewSaveConfig(options...)
+
+	// Compute ETag for the current state of the object.
+	raw, err := json.Marshal(obj.Data)
+	if err != nil {
+		return err
+	}
+
+	obj.ETag = etag.New(raw)
+
+	// We need different SQL for the case where an etag is provided vs not provided.
+	//
+	// The key behavior difference is that if an etag is provided, we should not perform inserts, only updates.
+
+	// This is the more complex query that handles "upserts". It does not process etags.
+	sql := `
+WITH updated AS (
+	INSERT INTO resources (id, original_id, resource_type, root_scope, routing_scope, etag, resource_data)
+	VALUES ($1, $2, $3, $4, $5, $6, $7)
+	ON CONFLICT (id) 
+	DO UPDATE SET resource_data = $7
+	RETURNING id
+)
+SELECT
+CASE
+	WHEN EXISTS (SELECT 1 FROM updated) THEN 'Success'
+	WHEN EXISTS (SELECT 1 FROM resources WHERE id = $1) THEN 'ErrConcurrency'
+	ELSE 'ErrNotFound'
+END AS result;`
+
+	args := []any{
+		storeutil.NormalizePart(converted.String()),
+		obj.ID, // MUST NOT BE NORMALIZED. Preserve the original casing and format.
+		storeutil.NormalizePart(converted.Type()),
+		storeutil.NormalizePart(converted.RootScope()),
+		storeutil.NormalizePart(converted.RoutingScope()),
+		obj.ETag,
+		obj.Data,
+	}
+
+	if config.ETag != "" {
+		// This is the simpler query that only performs updates. It requires an etag.
+		// NOTE: we want to report ErrConcurrency for all failure cases here. This is what the tests do.
+		sql = `
+WITH updated AS (
+	UPDATE resources SET resource_data = $2
+	WHERE id = $1 AND etag = $3
+	RETURNING id
+)
+SELECT
+CASE
+	WHEN EXISTS (SELECT 1 FROM updated) THEN 'Success'
+	WHEN EXISTS (SELECT 1 FROM resources WHERE id = $1) THEN 'ErrConcurrency'
+	ELSE 'ErrConcurrency'
+END AS result;`
+
+		args = []any{storeutil.NormalizePart(converted.String()), obj.Data, config.ETag}
+	}
+
+	result := ""
+	err = p.api.QueryRow(ctx, sql, args...).Scan(&result)
+	if err != nil {
+		return err
+	} else if result == "ErrNotFound" {
+		return &store.ErrNotFound{ID: obj.ID}
+	} else if result == "ErrConcurrency" {
+		return &store.ErrConcurrency{}
+	}
+
+	return nil
+}
+
+// createPaginationToken converts a timestamp to a base64 encoded string.
+//
+// We use ISO8601/RFC3339 format which postgres understands and can be used for comparison.
+// We also add microseconds to the timestamp to ensure uniqueness.
+func (p *PostgresClient) createPaginationToken(timestamp time.Time) (string, error) {
```

Comment:
Now I see that we don't need UUID for pagination.

Reply:
👍 The timestamp-cursor based approach is pretty well understood. I'd like to stick with that unless there's a good reason to change.